### ISSUE TITLE
Return Async from work()

### ISSUE THREAD

**Issue by liyinsg** [ClickHouse/ClickHouse#41757]
I can see below comment for IProcessor::schedule(), however only the second method is used.
```
      * NOTE: it would be more logical to let `work()` return ASYNC status instead of prepare. This will get
      * prepare() -> work() -> schedule() -> work() -> schedule() -> .. -> work() -> prepare()
      * chain instead of
      * prepare() -> work() -> prepare() -> schedule() -> work() -> prepare() -> schedule() -> .. -> work() -> prepare()
```
https://github.com/ClickHouse/ClickHouse/blob/e8978165722603b3f902430552abdc3453e096c6/src/Processors/IProcessor.h#L216

My question is, can we return Async from work(), or it's just an outdated comment?

**Comment by ucasfl** [comment]
The comment is saying it's better to let `work()` return ASYNC but current implmentation is not( return ASYNC by `prepare()`).

**Comment by liyinsg** [comment]
Thanks for the explanation, I guess we can even skip the call of schedule() too if we can return fd > 0 from work()?
```
prepare() -> work() -> ... -> work() -> prepare()
```

**Comment by ucasfl** [comment]
> Thanks for the explanation, I guess we can even skip the call of schedule() too if we can return fd > 0 from work()?
> 
> ```
> prepare() -> work() -> ... -> work() -> prepare()
> ```

Yes, but in current interface, the return type of `work()` is `void`.
